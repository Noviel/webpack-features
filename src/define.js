import webpack from 'webpack';

export default (
  { target, production },
  {
    __prefix = 'process.env.',
    NODE_ENV = production ? 'production' : 'development',
    ...rest
  } = {}
) => {
  const defines = {};

  if (NODE_ENV) {
    defines['process.env.NODE_ENV'] = JSON.stringify(NODE_ENV);
  }

  Object.keys(rest).forEach(key => {
    defines[`${__prefix}${key}`] = JSON.stringify(rest[key]);
  });

  return new webpack.DefinePlugin(defines);
};
