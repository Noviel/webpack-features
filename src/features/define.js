// @flow
import webpack from 'webpack';

import type { Exact, Env, PluginExtendOptions } from '../lib/types';

type _Options = {
  $prefix: string,
  NODE_ENV: string,
  [string]: any,
};

export type Options = Exact<_Options>;

export type Result = {|
  plugins: [webpack.DefinePlugin],
|};

export default (
  env: Env,
  {
    $prefix = 'process.env.',
    NODE_ENV = env.production ? 'production' : 'development',
    ...defines
  }: Options = {},
  { plugins, next }: PluginExtendOptions
): Result => {
  const definitions = {};

  if (NODE_ENV) {
    definitions['process.env.NODE_ENV'] = JSON.stringify(NODE_ENV);
  }

  Object.keys(defines).forEach((key: string) => {
    definitions[`${$prefix}${key}`] = JSON.stringify(defines[key]);
  });

  return next(env, plugins, {
    plugins: [new webpack.DefinePlugin(definitions)],
  });
};
