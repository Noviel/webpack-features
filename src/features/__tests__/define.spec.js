import define from '../define';

import { callFeature, envs } from '../../fixtures';

describe('define', () => {
  it('should set correct NODE_ENV based on the `env.production`', () => {
    expect(
      callFeature(define).plugins[0].definitions['process.env.NODE_ENV']
    ).toBe('"production"');
    expect(
      callFeature(define, {}, [], envs.modernBrowsersDev).plugins[0]
        .definitions['process.env.NODE_ENV']
    ).toBe('"development"');
  });

  it('should override NODE_ENV', () => {
    expect(
      callFeature(define, { NODE_ENV: 'TEST' }).plugins[0].definitions[
        'process.env.NODE_ENV'
      ]
    ).toBe('"TEST"');
  });

  it('should not define NODE_ENV if it set to `false`', () => {
    expect(
      callFeature(define, { NODE_ENV: false }).plugins[0].definitions
    ).not.toHaveProperty('process.env.NODE_ENV');
  });

  it('should add primitive values to definitions', () => {
    expect(
      callFeature(define, { VALUE: 100, SECOND: true, THIRD: 'THIRD' })
        .plugins[0].definitions
    ).toMatchSnapshot();
  });

  it('should set a prefix for the every definition except NODE_ENV', () => {
    expect(
      callFeature(define, { $prefix: 'custom_prefix.', VALUE: 500 }).plugins[0]
        .definitions
    ).toMatchSnapshot();
  });
});
