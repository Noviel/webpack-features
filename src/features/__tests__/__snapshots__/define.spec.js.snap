// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`define should add primitive values to definitions 1`] = `
Object {
  "process.env.NODE_ENV": "\\"production\\"",
  "process.env.SECOND": "true",
  "process.env.THIRD": "\\"THIRD\\"",
  "process.env.VALUE": "100",
}
`;

exports[`define should apply a plugin 1`] = `
Object {
  "plugins": Array [
    DefinePlugin {
      "definitions": Object {
        "process.env.NODE_ENV": "\\"production\\"",
      },
    },
  ],
  "target": Object {
    "name": "browsers",
    "value": "modern",
  },
}
`;

exports[`define should set a prefix for the every definition except NODE_ENV 1`] = `
Object {
  "custom_prefix.VALUE": "500",
  "process.env.NODE_ENV": "\\"production\\"",
}
`;
